bluetooth: controller: Revised ticker for improved conflict resolution

mtpr-ot opened the pull request:
Changed ticker behavior to resolve conflicts in ticker_worker instead of
in ticker_job. This allows better real-time slot allocation, as well as
allowing callbacks even if ticker node doesn't get air-time (for audio
stream housekeeping). A priority property now also allows prioritizing
one ticker node over others.

Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak reviewed (dismissed):
I need time until 7th May to do some manual tests and profiling.

Do you have any readily available CPU usages in typical usecases (connection event CPU use in comparison)?

mtpr-ot commented:
> I need time until 7th May to do some manual tests and profiling.
> 
> Do you have any readily available CPU usages in typical usecases (connection event CPU use in comparison)?

I don't have any CPU usage number. I only tested on target by means of checking packet loss in BLE- and audio connection scenarios, with simultaneous advertising or scanning.

mtpr-ot commented:
@cvinayak: Please take your time. It is important to make sure this works for all. If you can come up with a worst case scenario to test, I will be happy to do so with our platform.

cvinayak commented:
@mtpr-ot Just to let you know that I want to spend time on this PR, but have not been able to do this week. Basically, I want to profile (using GPIO pins) for CPU time on nRF51 and nRF52, under:
1. Typical peripheral sample
2. Typical central_hr sample
3. Peripheral with slave latencies.
4. Run conformance tests (atleast to ensure that change in CPU use has no regression)

Finger in the air, I am thinking to timebox 1 week of effort.

mtpr-ot commented:
> @mtpr-ot Just to let you know that I want to spend time on this PR, but have not been able to do this week. Basically, I want to profile (using GPIO pins) for CPU time on nRF51 and nRF52, under:
> 
> 1. Typical peripheral sample
> 2. Typical central_hr sample
> 3. Peripheral with slave latencies.
> 4. Run conformance tests (atleast to ensure that change in CPU use has no regression)
> 
> Finger in the air, I am thinking to timebox 1 week of effort.

Sounds great! Only, I hope it will not be an actual man-week of work, in which case I would say there are probably more important stuff on your list!

mped-oticon commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -329,93 +295,68 @@ static void ticker_by_next_slot_get(struct ticker_instance *instance,
-			node[current].ticks_to_expire -= ticks_to_expire;
-		}
+	if (previous == TICKER_NULL) {
+		instance->ticker_id_head = id;
```
Intent: Queue was empty (no previous), so set the new head.
However, it is possible that the very first loop through above, didn't set previous (because ticks_to_expire == 0 did a break before updating previous)

mtpr-ot replied:
ticks_to_expire cannot be 0

mped-oticon commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -127,6 +134,7 @@ struct ticker_user_op {
 		struct ticker_user_op_start start;
 		struct ticker_user_op_update update;
 		struct ticker_user_op_slot_get slot_get;
+		s8_t priority;	   /* Node priority. Defaults to 0 */
```
This is part of a union, which seems unclear - until one realizes it belongs to TICKER_USER_OP_TYPE_PRIORITY_SET for which we lack a user op struct. 
Would be nice to add that struct for consistency.

mtpr-ot replied:
Will create a struct for it.

mped-oticon commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -492,6 +433,87 @@ static u32_t ticker_dequeue(struct ticker_instance *instance, u8_t id)
+			}
+
+			u8_t force = (ticker_next->force > ticker->force);
+			u8_t has_priority =
```
Here is effectively our priority policy/metric. But I am not sure adding the latency bias is The Best For All. It would be good to keep it simple with just priority, and outsource the policy to App (by having it change the priority).

mped-oticon commented:
Looks good. I have a private branch with more commentary that I will share at some point.

A few explanatory notes:

- top-half:    worker - looks at head of timer queue, priority resolution, spawns payload callback, driven by timer.
- bottom-half: job    - process user-operation queue and updates timer queue.

--

With this PR, 
 - we no longer resolve conflicts ahead of time but resolve conflicts in the worker when needed, i.e. as late as possible.
    - The reason for this is due to causality and the dynamic behaviour of audio:
    - We can not know in advance if the next event must be scheduled; only when the current event is done will we know.
    - This implies the timer queue can now contain conflicts -- which is good as the following sequence of operations will return to status-quo: [insert of X] followed by [removal of X] where X may collide with any - number of other events.
    - This is exactly what we need for our proprietary protocols.
 - Introduce priority 
    - Currently, priority policy is part of the ticker, and is a mix of priority and latency.
    - However, we may want to outsource this policy: There is no One Best policy; it is application specific.
 - Introduce shallow expiration
    - Needed for audio, when no air-time was granted but we still have to maintain an active audio pipeline.

--

Now that conflict resolution has moved into worker, some (small) increase in start latency is to be expected. For pathological cases, this latency is O(n) time for n connections. This would have to be measured as @cvinayak mentions. I don't think it will be a problem. But in case it is, we can apply speculation: After every update to the queue, perform conflict resolution ahead of time but without letting conflict resolution modify the queue. This preserves all the benefits above while restoring the O(1)-time worker latency. 

-- 
Finally, the scheduling (both now and before) has been greedy. If we find QoS to be poor when we start stressing with many connections, it may be worth looking into Interval or EDF Scheduling:
https://en.wikipedia.org/wiki/Interval_scheduling
https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling

mtpr-ot commented:
Added fix from ot_master2 + TICKER_PRIORITY_CRITICAL handling. @mped-oticon, please review this part as well.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -492,6 +440,94 @@ static u32_t ticker_dequeue(struct ticker_instance *instance, u8_t id)
+			struct ticker_node *ticker_next = &nodes[id_head];
+
+			/* Accumulate ticks_to_expire for each node */
+			acc_ticks_to_expire += ticker_next->ticks_to_expire;
```
This is a fix!

mtpr-ot commented:
Tak Mark!

Jeg har dog lige taget konsekvensen af at det kan tage lang tid for os at få PR’er igennem nu, og har opdateret denne PR med en rettelse fra ot_master2 + tilføjelse af TICKER_PRIORITY_CRITICAL (win-all prioritet).
Håber det er OK at lige kigge på det også.

/Morten

From: Mark Ruvald Pedersen [mailto:notifications@github.com]
Sent: 10. juni 2019 16:55
To: zephyrproject-rtos/zephyr <zephyr@noreply.github.com>
Cc: Morten Priess (MTPR) <mtpr@demant.com>; Mention <mention@noreply.github.com>
Subject: Re: [zephyrproject-rtos/zephyr] bluetooth: controller: Revised ticker for improved conflict resolution (#15787)


Looks good. I have a private branch with more commentary that I will share at some point.

A few explanatory notes:

  *   top-half: worker - looks at head of timer queue, priority resolution, spawns payload callback, driven by timer.
  *   bottom-half: job - process user-operation queue and updates timer queue.

--

With this PR,

  *   we no longer resolve conflicts ahead of time but resolve conflicts in the worker when needed, i.e. as late as possible.
     *   The reason for this is due to causality and the dynamic behaviour of audio:
     *   We can not know in advance if the next event must be scheduled; only when the current event is done will we know.
     *   This implies the timer queue can now contain conflicts -- which is good as the following sequence of operations will return to status-quo: [insert of X] followed by [removal of X] where X may collide with any - number of other events.
     *   This is exactly what we need for our proprietary protocols.
  *   Introduce priority
     *   Currently, priority policy is part of the ticker, and is a mix of priority and latency.
     *   However, we may want to outsource this policy: There is no One Best policy; it is application specific.
  *   Introduce shallow expiration
     *   Needed for audio, when no air-time was granted but we still have to maintain an active audio pipeline.

--

Now that conflict resolution has moved into worker, some (small) increase in start latency is to be expected. For pathological cases, this latency is O(n) time for n connections. This would have to be measured as @cvinayak<https://github.com/cvinayak> mentions. I don't think it will be a problem. But in case it is, we can apply speculation: After every update to the queue, perform conflict resolution ahead of time but without letting conflict resolution modify the queue. This preserves all the benefits above while restoring the O(1)-time worker latency.

--
Finally, the scheduling (both now and before) has been greedy. If we find QoS to be poor when we start stressing with many connections, it may be worth looking into Interval or EDF Scheduling:
https://en.wikipedia.org/wiki/Interval_scheduling
https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/15787?email_source=notifications&email_token=AJRTOLS7QGXXB25NNHUGQC3PZZTMBA5CNFSM4HJTQKV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXKDNLA#issuecomment-500446892>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AJRTOLTE42AOUUI4UM264VDPZZTMBANCNFSM4HJTQKVQ>.

mtpr-ot committed:
bluetooth: controller: Revised ticker for improved conflict resolution

Changed ticker behavior to resolve conflicts in ticker_worker instead of
in ticker_job. This allows better real-time slot allocation, as well as
allowing callbacks even if ticker node doesn't get air-time (for audio
stream housekeeping). A priority property now also allows prioritizing
one ticker node over others.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mped-oticon commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1983,6 +1964,63 @@ u32_t ticker_job_idle_get(u8_t instance_index, u8_t user_id,
+ *			     Lover value equals higher priority. Setting
+ *			     priority to -128 (TICKER_PRIORITY_CRITICAL) makes
+ *			     the node win all collision challenges. Only one
+ *			     node can have this priority assigned.
```
What happens if >=2 nodes have critical priority?

mtpr-ot replied:
In that case it's first-come first-served. The critical priority is not guaranteed honored.

mped-oticon commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1606,6 +1583,10 @@ u32_t ticker_init(u8_t instance_index, u8_t count_node, void *node,
 
+	while (count_node--) {
+		instance->node[count_node].priority = 0;
+	}
```
This should be done in `ticker_job_op_start`. If not, the priorities are retained between ticker start/stop sessions, when they are different things.

mtpr-ot replied:
This is intentional. During conn interval updates, tickers are stopped and restarted in ULL. If priority is initialized in the start operation, priority is "suddenly" lost. In my opinion, letting priority stick makes sense because the application should always set priority on a connection anyway (if needed).

mped-oticon replied:
As me and Morten talked about offline, this priority "spill over" should be fixed. A possible solution would be to introduce a Restart user op - to be used by conn update. This leaves Start to do zero init of priority.

mtpr-ot replied:
> As me and Morten talked about offline, this priority "spill over" should be fixed. A possible solution would be to introduce a Restart user op - to be used by conn update. This leaves Start to do zero init of priority.

We decided to accept it as is, and create an issue on it.

aescolar commented:
Merging as agreed on Tuesday 2019/06/11

cvinayak commented:
[peripheral_sample_CPU_profile.zip](https://github.com/zephyrproject-rtos/zephyr/files/3285883/peripheral_sample_CPU_profile.zip)

Based on simple peripheral usecase CPU utilization comparison, this PR does not impact performance. That said, conflict resolution has not been compared.

Quick review, and no new comments from my end. This is a good work done.
